File: t-tasks/src/main/java/com/teo/ttasks/jobs/DeleteTaskJob.java
package com.teo.ttasks.jobs;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.birbit.android.jobqueue.CancelReason;
import com.birbit.android.jobqueue.Job;
import com.birbit.android.jobqueue.Params;
import com.birbit.android.jobqueue.RetryConstraint;
import com.google.firebase.database.DatabaseReference;
import com.teo.ttasks.api.TasksApi;
import com.teo.ttasks.data.model.TTask;
import com.teo.ttasks.util.FirebaseUtil;

import javax.inject.Inject;

import io.realm.Realm;
import retrofit2.Response;
import timber.log.Timber;

public class DeleteTaskJob extends Job {

    @Inject transient TasksApi tasksApi;

    private String taskId;
    private String taskListId;

    public DeleteTaskJob(String taskId, String taskListId) {
        super(new Params(Priority.HIGH).requireNetwork().persist());
        this.taskId = taskId;
        this.taskListId = taskListId;
    }

    @Override
    public void onAdded() {
        // Do nothing
    }

    public void onRun() throws Throwable {
        // Delete the reminder
        final DatabaseReference tasksDatabase = FirebaseUtil.getTasksDatabase();
        FirebaseUtil.saveReminder(tasksDatabase, taskId, null);

        final Realm realm = Realm.getDefaultInstance();
        final TTask tTask = realm.where(TTask.class).equalTo("id", taskId).findFirst();

        // Task not found, nothing to do here
        if (tTask == null) {
            realm.close();
            return;
        }

        // Delete the Google task
        if (!tTask.isLocalOnly()) {
            final Response<Void> response = tasksApi.deleteTask(taskListId, taskId).execute();
            response.body();

            // Handle failure
            if (!response.isSuccessful()) {
                realm.close();
                response.errorBody().close();
                throw new Exception("Failed to delete task");
            }
        }

        // Delete the Realm task
        realm.executeTransaction(realm1 -> TTask.deleteFromRealm(tTask));
        Timber.d("deleted task %s", taskId);

        realm.close();
    }

    @Override
    protected void onCancel(@CancelReason int cancelReason, @Nullable Throwable throwable) {
        if (throwable != null)
            Timber.e(throwable.toString());
    }

    @Override
    protected RetryConstraint shouldReRunOnThrowable(@NonNull Throwable throwable, int runCount, int maxRunCount) {
        return RetryConstraint.createExponentialBackoff(runCount, 1000);
    }
}

File: t-tasks/src/main/java/com/teo/ttasks/jobs/DeleteTaskJob.kt
package com.teo.ttasks.jobs

import com.evernote.android.job.Job
import com.evernote.android.job.JobManager
import com.evernote.android.job.JobRequest
import com.evernote.android.job.util.support.PersistableBundleCompat
import com.google.firebase.database.FirebaseDatabase
import com.teo.ttasks.TTasksApp
import com.teo.ttasks.api.TasksApi
import com.teo.ttasks.data.remote.TasksHelper
import com.teo.ttasks.delete
import com.teo.ttasks.util.FirebaseUtil.getTasksDatabase
import com.teo.ttasks.util.FirebaseUtil.saveReminder
import io.realm.Realm
import timber.log.Timber
import java.util.concurrent.TimeUnit
import javax.inject.Inject

class DeleteTaskJob : Job() {

    companion object {
        const val TAG = "DELETE_TASK"
        const val EXTRA_TASK_ID = "taskId"
        const val EXTRA_TASK_LIST_ID = "taskListId"

        fun schedule(taskId: String, taskListId: String) {
            JobManager.instance().getAllJobRequestsForTag(TAG)
                    .forEach {
                        if (it.extras[EXTRA_TASK_ID] as String == taskId &&
                                it.extras[EXTRA_TASK_LIST_ID] as String == taskListId) {
                            Timber.v("Delete Task job already exists for %s, ignoring...", taskId)
                            return
                        }
                    }
            val extras = PersistableBundleCompat().apply {
                putString(EXTRA_TASK_ID, taskId)
                putString(EXTRA_TASK_LIST_ID, taskListId)
            }
            JobRequest.Builder(TAG)
                    .setBackoffCriteria(5_000L, JobRequest.BackoffPolicy.EXPONENTIAL)
                    .setRequiredNetworkType(JobRequest.NetworkType.CONNECTED)
                    .setExtras(extras)
                    .setRequirementsEnforced(true)
                    .setUpdateCurrent(true)
                    .setExecutionWindow(1, TimeUnit.DAYS.toMillis(1))
                    .build()
                    .schedule()
        }
    }

    @Inject @Transient internal lateinit var tasksHelper: TasksHelper
    @Inject @Transient internal lateinit var tasksApi: TasksApi

    override fun onRunJob(params: Params): Result {
        // Inject dependencies
        (context.applicationContext as TTasksApp).applicationComponent().inject(this)

        // Extract params
        val extras = params.extras
        val taskId = extras[EXTRA_TASK_ID] as String
        val taskListId = extras[EXTRA_TASK_LIST_ID] as String

        // Delete the reminder
        val tasksDatabase = FirebaseDatabase.getInstance().getTasksDatabase()
        tasksDatabase.saveReminder(taskId, null)

        val realm = Realm.getDefaultInstance()
        val tTask = tasksHelper.getTask(taskId, realm)

        // Task not found, nothing to do here
        if (tTask == null) {
            realm.close()
            return Job.Result.SUCCESS
        }

        // Delete the Google task
        if (!tTask.isLocalOnly) {
            val result = tasksApi.deleteTask(taskListId, taskId).blockingGet()
            if (result != null) {
                Timber.e(result, "Error while deleting remote task")
                realm.close()
                return Job.Result.RESCHEDULE
            }
        }

        // Delete the Realm task
        realm.executeTransaction { tTask.delete() }
        Timber.v("Deleted task %s", taskId)

        realm.close()
        return Job.Result.SUCCESS
    }

    override fun onCancel() {
        Timber.e("Delete Task job cancelled")
//        throwable?.let { Timber.e(it.toString()) }
    }
}
